math/vecops/ARCHITECTURE.md

```diff
@@ -0,0 +1,103 @@
+# Architecture of the VecOps library
+
+> This document is meant for ROOT developers, to quickly get their bearings around the VecOps library.
+
+The main type in the library is `RVec`. Besides `RVec`, the library only contains helper types and functions.
+
+`RVec` is a vector type that tries to be as `std::vector`-like as possible while adding a
+few important features, namely:
+- the ability to act as a view over an existing memory buffer (see "Memory adoption" below)
+- a small-buffer optimization
+- vectorized operator overloads
+- a vectorized `operator[](mask)` to allow quick element selection together with vectorized operators
+  (e.g. `etas[etas > k]` returns a new `RVec` with all elements greater than `k`)
+- helper functions such as `InvariantMass`, `DeltaR`, `Argsort` are also provided
+
+The current implementation of `RVec` is based on LLVM's SmallVector, extracted
+from the head of LLVM's repo around December 2020.
+We are not tracking the upstream implementation.
+
+Compared to LLVM's SmallVectors:
+
+- memory adoption capabilities have been added
+- patches have been applied to make RVec work with (ROOT's version of) cppyy (notably `using` declarations had to be
+  lowered in the inheritance hierarchy for cppyy to pick them up)
+- `operator[](mask)` has been added, as well as several other "numpy-like" helper
+  functions (these latter ones are free functions)
+- logical operators `==`, `<`, `>` etc. return vectors rather than booleans
+- the type of fSize and fCapacity is signed rather than unsigned, and fixed to 32 bits
```

Comment:
@enrico Maybe that has been already discussed, but are we sure that we'll never support `RVec` > 2G?

Reply:
`fSize` is the number of elements, not the size in bytes of the object. we don't expect users to ever instantiate RVecs with size larger than 2^31 . what would the usecase be?

Reply:
Can we detect the case and throw?

Reply:
We already do, see `grow` and `grow_pod`